### ISSUE TITLE
"Cannot find type" when trying to create PSADKeyCredential object

### ISSUE THREAD

**Issue by janus-christensen** [Azure/azure-powershell#4491]
### Cmdlet(s)
New-Object

### PowerShell Version
PSVersion                      5.1.14393.206 

### OS Version
BuildVersion                   10.0.14393.206

### Description
I am attempting to create a new PSADKeyCredential object using New-Object, but I am running into an issue where I get the error "Cannot find type [Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.PSADKeyCredential]"

When I write specify the type in Powershell with command completion, it looks like the *.Models.ActiveDirectory subtree is missing.

Is there something I need to import besides AzureRM.Resources?

Thank you,
Janus

### Debug Output
    New-Object : Cannot find type [Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.PSADKeyCredential]: verify that the assembly containing this type is loaded.
    At line:1 char:18
    + ... redential = New-Object Microsoft.Azure.Commands.Resources.Models.Acti ...
    +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidType: (:) [New-Object], PSArgumentException
        + FullyQualifiedErrorId : TypeNotFound,Microsoft.PowerShell.Commands.NewObjectCommand

### Script/Steps for Reproduction

    Import-Module AzureRM.Resources
    $keyCredential = New-Object Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.PSADKeyCredential

**Comment by Welasco** [comment]
I have the same problem.
The only available workaround so far that I found is downgrading the powershell version to 4.2.1.

Install-Module AzureRM -RequiredVersion 4.2.1

Something happen after this release.

**Comment by sphibbs** [comment]
@cormacpayne please investigate

**Comment by MaxFrost** [comment]
I think I can shed some light on this. Commit https://github.com/Azure/azure-powershell/commit/fdddee667832e720213c57a6d976c474ecde9a10#diff-ae5a686d2a2de07a6f1f57fd033f7777 appears to have moved them into a different area, no longer under resources..models.activedirectory, they are now under Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.

so if you were using Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.PSADKeyCredential before you would now use Miicrosoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.PSADPasswordCredential

**Comment by MaxFrost** [comment]
So knowing that, and with some help from @jaykul, sourced that we should _not_ be using this new-object, but instead be using New-AzureRmADAppCredential to generate the credential, as it's output is our PSADKeyCredential that we're looking for.

**Comment by Jaykul** [comment]
@cormacpayne I'd close this. It's as-designed.

@Welasco @janus-christensen I think the intention (of this module) is that you should not use the underlying classes directly -- the team considers them a convenience for the module developers, and not an interface for users...

**Comment by wilddev65** [comment]
Maybe I'm misunderstanding this, but the following page talks about using the PSADKeyCredential to setup a service principal (which is replicated in webpages all over the net).
https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal
The script used in that example uses New-AzureRMADServicePrincipal with the PSADKeyCredential setup by the New-Object command that no longer works. I'm confused about how you would use New-AzureRmADAppCredential when setting up the service principal instead.

**Comment by MaxFrost** [comment]
Well, you would use New-AzureRmADSpCredential instead. Create your Service Principal without a password or any credentials (so no keycredentials, , etc), then immediately run the New-AzureRmADSpCredential right after with the values you would have supplied to -KeyCredentials.

Using the Class only makes sense if the class would always remain, but if they moved it once, it'll likely get moved again, and the documentation should be updated to show the new usage.

**Comment by cormacpayne** [comment]
@Jaykul that is correct, Joel! The Azure PowerShell team views these objects (returned from cmdlets) as PowerShell objects and not .NET objects - we are only concerned about the properties of a type, and not about its methods, constructor, etc. We do not intend to have objects of these types created using `New-Object`, so the documentation that @wilddev65 linked to is incorrect and will need to be updated (thanks for pointing that out!).

@sdwheeler Hey Sean, would you mind checking out the documentation linked above by Chris ( @wilddev65 ) and see if we can remove the `New-Object` call that is creating an object of type `Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.PSADKeyCredential` ? Feel free to ping us if you have any questions.

Closing as this is by-design.

**Comment by sdwheeler** [comment]
@cormacpayne @tfitzmac is the author of the article. Tom can you remove the "New-Object" call from your article?

**Comment by tfitzmac** [comment]
Thanks for this info, I am working on a fix right now.

**Comment by janus-christensen** [comment]
The article that @wilddev65 links to is one of the articles I  originally used to figure out how to do this. I will look into how to use New-AzureRmADAppCredential. 

Thank you for all the responses and help.
/Janus

**Comment by Welasco** [comment]
Hi Guys, 
I found this doc that need's update as well:
https://docs.microsoft.com/en-us/azure/automation/automation-update-account-powershell

**Comment by yobyot** [comment]
Me, too. Same problem: trying to work through the PowerShell example to add a credential and stuck on not being able to run the sample code. BTW, where _are_ the objects documented? I couldn't find them easily. FWIW, I am used to AWS where objects are linked from the cmdlet doc. Wish Azure would do the same.

**Comment by nbmsm1** [comment]
I have observed that password credentials created with **PSADKeyCredential** show up when viewing the application in the portal, whereas password credentials created with **New-AzureRmADSpCredential** do not. Is this correct?

**Comment by Obulab** [comment]
Thanks , providing  certvalue directly with New-AzureRMADServicePrincipal worked for me.

**Comment by Zazcallabah** [comment]
If we aren't supposed to use PSADKeyCredental objects, how are we supposed to call New-AzureRmADApplication with the ApplicationWithPasswordCredentialParameterSet?

**Comment by pixelicous** [comment]
Classic, breaking automation and tools because of upgrades, all documentation still points to non working code.

**Comment by winmike** [comment]
I have same issue even w following modules in stack loaded - still "type not found"

Script     3.1.0      azure.storage
Script     3.1.0      azurerm.automation
Script     3.1.0      azurerm.keyvault
Script     3.1.0      azurerm.profile
Script     4.1.0      azurerm.resources
Script     3.1.0      azurerm.sql
Script     3.1.0      azurerm.storage
Manifest   3.1.0.0    Microsoft.PowerShell.Management
Manifest   3.0.0.0    Microsoft.PowerShell.Security
Manifest   3.1.0.0    Microsoft.PowerShell.Utility
Manifest   3.0.0.0    Microsoft.WSMan.Management
Manifest   1.1.166.0  msonline
Script     1.1.7.0    PackageManagement
Manifest   1.0.0.0    PKI
Script     1.6.0      PowerShellGet

**Comment by winmike** [comment]
SOLVED AS:
use 
New-Object Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.PSADKeyCredential

**Comment by PinakiIT** [comment]
I get this, using the code abouve, and then calling `New-AzureRmADApplication`
`Cannot convert the "Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.PSADKeyCredential" value of type "Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.PSADKeyCredential" to type "Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.PSADPasswordCredential`
